recipes/msys2/all/conanfile.py

```diff
@@ -11,54 +12,164 @@ class MSYS2Conan(ConanFile):
+                                 "http://repo.msys2.org/msys/i686/pacman-5.2.1-6-i686.pkg.tar.xz"]
+
+            if (self.settings.arch == "x86"):                     
+                for i in pacman_packages_x32:
+                    self.run('bash -l -c "pacman --noconfirm -U %s"' % i)
+
+            self._kill_pacman()
+
+
+    def _update_pacman(self):
+        with tools.chdir(os.path.join(self._msys_dir, "usr", "bin")):
+            self._update_very_old_pacman()
+
+            self._kill_pacman()
+            # https://www.msys2.org/news/   see  2020-05-31 - Update may fail with "could not open file"
+            # update pacman separately first
+            self.run('bash -l -c "pacman --noconfirm -Sydd pacman"')
+  
+            # https://www.msys2.org/docs/ci/
+            self.run('bash -l -c "pacman --noconfirm --ask 20 -Syuu"')  # Core update (in case any core packages are outdated)
+            self._kill_pacman()
+            self.run('bash -l -c "pacman --noconfirm --ask 20 -Syuu"')  # Normal update
+            self._kill_pacman()
+            self.run('bash -l -c "pacman -Rc dash --noconfirm"')
+
+    # https://github.com/msys2/MSYS2-packages/issues/1966
+    def _kill_pacman(self):
+        if (self.settings.os == "Windows"):
+            taskkill_exe = os.path.join(os.environ.get('SystemRoot'), 'system32', 'taskkill.exe')
+
+            log_out = True
+            if log_out:
+                out = subprocess.PIPE
+                err = subprocess.STDOUT
+            else:
+                out = file(os.devnull, 'w')
+                err = subprocess.PIPE
+
+            if os.path.exists(taskkill_exe):
+                taskkill_cmd1 = taskkill_exe + " /f /t /im pacman.exe"
+                taskkill_cmd2 = taskkill_exe + " /f /im gpg-agent.exe"
+                taskkill_cmd3 = taskkill_exe + " /f /im dirmngr.exe"
+
+                taskkill_cmd4 = taskkill_exe + ' /fi "MODULES eq msys-2.0.dll"'
+
+                try:
+                    proc = subprocess.Popen(taskkill_cmd1, stdout=out, stderr=err, bufsize=1)
+                    proc = subprocess.Popen(taskkill_cmd2, stdout=out, stderr=err, bufsize=1)
+                    proc = subprocess.Popen(taskkill_cmd3, stdout=out, stderr=err, bufsize=1)
+
+                    proc = subprocess.Popen(taskkill_cmd4, stdout=out, stderr=err, bufsize=1)
+                except OSError as e:
+                    if e.errno == errno.ENOENT:
+                        raise ConanException("Cannot kill pacman")
+
     @property
     def _msys_dir(self):
-        return "msys64" if self.settings.arch_build == "x86_64" else "msys32"
+        return "msys64" if (tools.Version(self.version) >= "20210105" or self.settings.arch == "x86_64") else "msys32"
 
-    def build(self):
-        arch = 0 if self.settings.arch_build == "x86" else 1  # index in the sources list
+    def source(self):
+        arch = 1 if self.settings.arch == "x86" else 0  # index in the sources list
```

Comment:
Do not use `settings` (or `options`) in the `source()` method. These sources are shared for all the configurations when working with Conan locally, if we need to retrieve different files depending on the configuration, ATM, they must be moved to `build()` method.

Conan v2. won't provide `self.settings` or `self.options` to the `sources()` method.